Defining more constants and deleting a function

smartie2076 opened the pull request:
Fix #266 

**Changes proposed in this pull request**:
- Added global variables to `constants_json_str.csv`, defining units with them 
- Removed function C0.complete_missing_cost_data() as this should be covered by A1 for csv files
- Added a setup_class (remove dir) to `test_B0.TestTemporaryJsonFileDisposal` 
- Installation help for `pygraphviz` on Win10/64bit systems in `troubleshooting.rst` 

The following steps were realized, as well (if applies):
- [x] Use in-line comments to explain your code
- [x] Write docstrings to your code
- [x] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code
- [x] Update the CHANGELOG.md
- [x] Apply black (`black . --exclude docs/`)
- [ ] Check if benchmark tests pass locally (`EXECUTE_TESTS_ON=master pytest`)

smartie2076 commented:
It was planned to solve #186 with this, but there are more things to consider to solve that issue.

Bachibouzouk commented on src/E2_economics.py:
```diff
@@ -3,6 +3,7 @@
 from src.constants_json_strings import (
```
git META tip: this could be a good use of `git commit --amend` (because you have 4 commits with very similar commit message one after the other)

smartie2076 replied:
Yeaaaaahhh... I know there was something to spare you all the mails. Sorry about that. I am never using the terminal for git, but always the gui, so I foget about that stuff...

Bachibouzouk replied:
I use both, gui for commits, then terminal for amend and rebase :)

Bachibouzouk commented:
What is left to do to merge this PR, fix the failing tests?

smartie2076 commented:
> 
> 
> What is left to do to merge this PR, fix the failing tests?

Exactly, there are some failing tests that as far as I know don´t really relate to what I changed.

smartie2076 commented:
Still some failures and warnings on my Win10 machine:

`
====================================================== 14 failed, 181 passed, 7 skipped, 20 warnings, 6 errors in 37.48s =======================================================`

smartie2076 commented:
Hi @Bachibouzouk ! So... running `pytests tests` on my win10 machine is a nightmare, and there are ALOT of error messages left to deciper, even when online the tests run through. Most are `FileExistsError`. I hope that adding "-f" to all `parse_args `in the `tests `relieves that (but honestly, this did not work for A0 even).

 Do you think adding the function `PARSE_ARGS` is something that introduces a double structure? Basically, we might be able to get around them if we change the the default values of the parser to `-log warning` and `-f` (at least for the tests). This will help to solve #312. 

@SabineHaas I am running into a couple of issues with D1 tests on my Win10 machine. I am a bit hesistant to try and fix them, both because I do not know how and because I do not want to mess with your tests...
Here they are:

```
(D:\PycharmProjects\mvs_eland\venv_windows) D:\PycharmProjects\mvs_eland>pytest tests/tests_D1_model_components.py
============================================================================= test session starts ==============================================================================
platform win32 -- Python 3.7.6, pytest-5.3.1, py-1.8.1, pluggy-0.13.1
rootdir: D:\PycharmProjects\mvs_eland
plugins: dash-1.11.0
collected 0 items                                                                                                                                                               

============================================================================ no tests ran in 0.02s =============================================================================
ERROR: file not found: tests/tests_D1_model_components.py


(D:\PycharmProjects\mvs_eland\venv_windows) D:\PycharmProjects\mvs_eland>pytest tests/tests_D1_model_components
============================================================================= test session starts ==============================================================================
platform win32 -- Python 3.7.6, pytest-5.3.1, py-1.8.1, pluggy-0.13.1
rootdir: D:\PycharmProjects\mvs_eland
plugins: dash-1.11.0
collected 0 items                                                                                                                                                               

============================================================================ no tests ran in 0.02s =============================================================================
ERROR: file not found: tests/tests_D1_model_components


(D:\PycharmProjects\mvs_eland\venv_windows) D:\PycharmProjects\mvs_eland>pytest tests/test_D1_model_components.py
============================================================================= test session starts ==============================================================================
platform win32 -- Python 3.7.6, pytest-5.3.1, py-1.8.1, pluggy-0.13.1
rootdir: D:\PycharmProjects\mvs_eland
plugins: dash-1.11.0
collected 27 items                                                                                                                                                              

tests\test_D1_model_components.py ..................FFFF.....                                                                                                             [100%]

=================================================================================== FAILURES ===================================================================================
_________________________________________________ TestSourceComponent.test_source_dispatchable_optimize_normalized_timeseries __________________________________________________

self = <mvs_eland.tests.test_D1_model_components.TestSourceComponent object at 0x000001F6C6658408>

    def test_source_dispatchable_optimize_normalized_timeseries(self):
        dict_asset = self.dict_values[ENERGY_PRODUCTION]["dispatchable_source_optimize"]
        dict_asset[TIMESERIES_NORMALIZED] = self.time_series / max(self.time_series)
        dict_asset[TIMESERIES_PEAK] = {"unit": "kWp/H", "value": self.time_series.max()}
    
        D1.source(
            model=self.model,
            dict_asset=dict_asset,
            sources=self.sources,
            busses=self.busses,
        )
    
        # checks done with helper function (see func for more information)
        self.helper_test_source_in_model_and_dict(
            dict_asset=dict_asset,
            dispatchable=True,
            mode="optimize",
>           timeseries="normalized",
        )

tests\test_D1_model_components.py:500:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests\test_D1_model_components.py:423: in helper_test_source_in_model_and_dict
    output_bus.max, dict_asset[TIMESERIES_NORMALIZED]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

left = [], right = 0    0.833333
1    0.916667
2    1.000000
dtype: float64, cls = <class 'pandas.core.series.Series'>

    def _check_isinstance(left, right, cls):
        """
        Helper method for our assert_* methods that ensures that
        the two objects being compared have the right type before
        proceeding with the comparison.
    
        Parameters
        ----------
        left : The first object being compared.
        right : The second object being compared.
        cls : The class type to check against.
    
        Raises
        ------
        AssertionError : Either `left` or `right` is not an instance of `cls`.
        """
    
        err_msg = "{name} Expected type {exp_type}, found {act_type} instead"
        cls_name = cls.__name__
    
        if not isinstance(left, cls):
            raise AssertionError(err_msg.format(name=cls_name, exp_type=cls,
>                                               act_type=type(left)))
E           AssertionError: Series Expected type <class 'pandas.core.series.Series'>, found <class 'oemof.solph.plumbing._Sequence'> instead

venv_windows\lib\site-packages\pandas\util\testing.py:353: AssertionError
__________________________________________ TestSourceComponent.test_source_dispatchable_optimize_timeseries_not_normalized_timeseries __________________________________________

self = <mvs_eland.tests.test_D1_model_components.TestSourceComponent object at 0x000001F6C666F408>

    def test_source_dispatchable_optimize_timeseries_not_normalized_timeseries(self,):
        dict_asset = self.dict_values[ENERGY_PRODUCTION]["dispatchable_source_optimize"]
        dict_asset[TIMESERIES] = self.time_series
        dict_asset[TIMESERIES_PEAK] = {"unit": "kWp/H", "value": self.time_series.max()}
    
        D1.source(
            model=self.model,
            dict_asset=dict_asset,
            sources=self.sources,
>           busses=self.busses,
        )

tests\test_D1_model_components.py:512:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src\D1_model_components.py:235: in source
    **kwargs,
src\D1_model_components.py:288: in check_optimize_cap
    func_optimize(model, dict_asset, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

model = <oemof.solph.network.EnergySystem object at 0x000001F6C6691908>
dict_asset = {'dispatch_price': {'unit': 'currency/kWh', 'value': 0.1}, 'dispatchable': True, 'installedCap': {'unit': 'kWp', 'value': 0}, 'label': 'Dispatchable source optimize
', ...}
kwargs = {'busses': {'Coal bus': "<oemof.solph.network.Bus: 'Coal bus'>", 'Electricity bus': "<oemof.solph.network.Bus: 'Electr...mof.solph.network.Bus: 'Electricity bus 2'>", '
Fuel bus': "<oemof.solph.network.Bus: 'Fuel bus'>", ...}, 'sources': {}}

    def source_non_dispatchable_optimize(model, dict_asset, **kwargs):
        r"""
        Defines a non dispatchable source with a capacity to be optimized.
    
        See :py:func:`~.source` for more information, including parameters.
    
        Returns
        -------
        Indirectly updated `model` and dict of asset in `kwargs` with the source object.
    
        """
        outputs = {
            kwargs["busses"][dict_asset[OUTPUT_BUS_NAME]]: solph.Flow(
                label=dict_asset[LABEL],
>               actual_value=dict_asset[TIMESERIES_NORMALIZED],
                fixed=True,
                investment=solph.Investment(
                    ep_costs=dict_asset[SIMULATION_ANNUITY][VALUE]
                    / dict_asset[TIMESERIES_PEAK][VALUE],
                    maximum=dict_asset[MAXIMUM_CAP][VALUE],
                    existing=dict_asset[INSTALLED_CAP][VALUE],
                ),
                # variable_costs are devided by time series peak as normalized time series are used as actual_value
                variable_costs=dict_asset[DISPATCH_PRICE][VALUE]
                / dict_asset[TIMESERIES_PEAK][VALUE],
            )
        }
E       KeyError: 'timeseries_normalized'

src\D1_model_components.py:634: KeyError
____________________________________________________ TestSourceComponent.test_source_dispatchable_fix_normalized_timeseries ____________________________________________________

self = <mvs_eland.tests.test_D1_model_components.TestSourceComponent object at 0x000001F6C66B0F88>

    def test_source_dispatchable_fix_normalized_timeseries(self):
        dict_asset = self.dict_values[ENERGY_PRODUCTION]["dispatchable_source_fix"]
        dict_asset[TIMESERIES_NORMALIZED] = self.time_series / max(self.time_series)
        dict_asset[TIMESERIES_PEAK] = {"unit": "kWp/H", "value": self.time_series.max()}
    
        D1.source(
            model=self.model,
            dict_asset=dict_asset,
            sources=self.sources,
>           busses=self.busses,
        )

tests\test_D1_model_components.py:531:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src\D1_model_components.py:235: in source
    **kwargs,
src\D1_model_components.py:280: in check_optimize_cap
    func_constant(model, dict_asset, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

model = <oemof.solph.network.EnergySystem object at 0x000001F6A60AD808>
dict_asset = {'dispatch_price': {'unit': 'currency/kWh', 'value': 0.1}, 'dispatchable': True, 'installedCap': {'unit': 'kWp', 'value': 0}, 'label': 'Dispatchable source fix', ..
.}
kwargs = {'busses': {'Coal bus': "<oemof.solph.network.Bus: 'Coal bus'>", 'Electricity bus': "<oemof.solph.network.Bus: 'Electr...mof.solph.network.Bus: 'Electricity bus 2'>", '
Fuel bus': "<oemof.solph.network.Bus: 'Fuel bus'>", ...}, 'sources': {}}

    def source_non_dispatchable_fix(model, dict_asset, **kwargs):
        r"""
        Defines a non dispatchable source with a fixed capacity.
    
        See :py:func:`~.source` for more information, including parameters.
    
        Returns
        -------
        Indirectly updated `model` and dict of asset in `kwargs` with the source object.
    
        """
        outputs = {
            kwargs["busses"][dict_asset[OUTPUT_BUS_NAME]]: solph.Flow(
                label=dict_asset[LABEL],
>               actual_value=dict_asset[TIMESERIES],
                fixed=True,
                nominal_value=dict_asset[INSTALLED_CAP][VALUE],
                variable_costs=dict_asset[DISPATCH_PRICE][VALUE],
            )
        }
E       KeyError: 'timeseries'

src\D1_model_components.py:606: KeyError
____________________________________________ TestSourceComponent.test_source_dispatchable_fix_timeseries_not_normalized_timeseries _____________________________________________

self = <mvs_eland.tests.test_D1_model_components.TestSourceComponent object at 0x000001F6C66B0C88>

    def test_source_dispatchable_fix_timeseries_not_normalized_timeseries(self):
        dict_asset = self.dict_values[ENERGY_PRODUCTION]["dispatchable_source_fix"]
        dict_asset[TIMESERIES] = self.time_series
        dict_asset[TIMESERIES_PEAK] = {"unit": "kWp/H", "value": self.time_series.max()}
    
        D1.source(
            model=self.model,
            dict_asset=dict_asset,
            sources=self.sources,
            busses=self.busses,
        )
        # checks done with helper function (see func for more information)
        self.helper_test_source_in_model_and_dict(
            dict_asset=dict_asset,
            dispatchable=True,
            mode="fix",
>           timeseries="not_normalized",
        )

tests\test_D1_model_components.py:557:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <mvs_eland.tests.test_D1_model_components.TestSourceComponent object at 0x000001F6C66B0C88>
dict_asset = {'dispatch_price': {'unit': 'currency/kWh', 'value': 0.1}, 'dispatchable': True, 'installedCap': {'unit': 'kWp', 'value': 0}, 'label': 'Dispatchable source fix', ..
.}
dispatchable = True, mode = 'fix', timeseries = 'not_normalized'

    def helper_test_source_in_model_and_dict(
        self, dict_asset, dispatchable, mode, timeseries=None
    ):
        """
        Helps testing whether `self.sources` and `self.model` was updated.
    
        Checks done:
        * self.sources contains the source (key = label, value = source object)
        * self.models contains the source (indirectly tested)
        * output bus has appropriate values for `actual_value`, investment and `variable_costs` (depending on `dispatchable`, `mode` and `timeseries`)
        * source only has one output flow
    
        """
        # self.sinks should contain the sink (key = label, value = sink object)
        assert dict_asset[LABEL] in self.sources
        assert isinstance(self.sources[dict_asset[LABEL]], solph.network.Source)
    
        # check amount of outputs from source (only one)
        assert len([str(i) for i in self.model.entities[-1].outputs]) == 1
    
        # self.models should contain the source (indirectly tested)
        # check output bus (`actual_value`, `investment` and `variable_costs`).
        # these values are expected to be different depending on `dispatchable`, `mode` and `timeseries`
        output_bus = self.model.entities[-1].outputs[
            self.busses[dict_asset[OUTPUT_BUS_NAME]]
        ]
        if mode == "fix":
            assert (
                output_bus.variable_costs.default == dict_asset[DISPATCH_PRICE][VALUE]
            )
            assert output_bus.investment is None
            if dispatchable == False:
                assert output_bus.nominal_value == dict_asset[INSTALLED_CAP][VALUE]
                assert_series_equal(output_bus.actual_value, dict_asset[TIMESERIES])
                assert output_bus.max == []
            elif dispatchable == True:
>               assert output_bus.existing == dict_asset[INSTALLED_CAP][VALUE]
E               AttributeError: 'Flow' object has no attribute 'existing'

tests\test_D1_model_components.py:402: AttributeError
=============================================================================== warnings summary ===============================================================================
venv_windows\lib\site-packages\oemof\network.py:15
venv_windows\lib\site-packages\oemof\network.py:15
  d:\pycharmprojects\mvs_eland\venv_windows\lib\site-packages\oemof\network.py:15: DeprecationWarning:

  Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working

venv_windows\lib\site-packages\pyutilib\misc\import_file.py:11
  d:\pycharmprojects\mvs_eland\venv_windows\lib\site-packages\pyutilib\misc\import_file.py:11: DeprecationWarning:

  the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=================================================================== 4 failed, 23 passed, 3 warnings in 3.64s ===================================================================

```

SabineHaas commented on src/constants_json_strings.py:
```diff
@@ -138,7 +143,20 @@
+# Autogenerated assets
+AUTO_SOURCE = "_source"
+AUTO_SINK = "_sink"
+DISPATCHABILITY = "is_dispatchable"
```
```suggestion
DISPATCHABILITY = "dispatchable"
```

this should solve the failing tests of D1

smartie2076 replied:
@SabineHaas niiiice! Thanks!

Bachibouzouk commented:
This can be merged soon right?

smartie2076 commented:
> 
> 
> This can be merged soon right?

Yes, it can. There is the thing about the `PARSE_ARGS()` that I introduced in `tests/constants.py` in an attempt to eliminate the windows test failures. I think by now that this is not needed, as it can already be set by using the `main()` function with arguments instead. Maybe we need to be able to have different default execution values for the tests, though.

Bachibouzouk commented:
> > This can be merged soon right?
> 
> Yes, it can. There is the thing about the `PARSE_ARGS()` that I introduced in `tests/constants.py` in an attempt to eliminate the windows test failures. I think by now that this is not needed, as it can already be set by using the `main()` function with arguments instead. Maybe we need to be able to have different default execution values for the tests, though.

I think making a new parser might create potential problems with default values (if we use multiple parsers, the already not explicit information of the default values, could be even more confusing and lead to test bugs because one could forget that the parser for the tests had this argument set to this value instead of that value.

Being explicit about the args passed as mock_arguments could not hurt. I would not feel bad if there is redundant test code (tests are not meant to be beautiful but effective and easy to understand.

smartie2076 commented:
> 
> 
> > > This can be merged soon right?
> > 
> > 
> > Yes, it can. There is the thing about the `PARSE_ARGS()` that I introduced in `tests/constants.py` in an attempt to eliminate the windows test failures. I think by now that this is not needed, as it can already be set by using the `main()` function with arguments instead. Maybe we need to be able to have different default execution values for the tests, though.
> 
> I think making a new parser might create potential problems with default values (if we use multiple parsers, the already not explicit information of the default values, could be even more confusing and lead to test bugs because one could forget that the parser for the tests had this argument set to this value instead of that value.
> 
> Being explicit about the args passed as mock_arguments could not hurt. I would not feel bad if there is redundant test code (tests are not meant to be beautiful but effective and easy to understand.

Okay, in that case I would remove the PARSE_ARGS again. It anyway did not solve our problems.

What do you think of having different default values for the MVS is pytest mode? Specifically I mean "-f" and "-log warning".

Bachibouzouk commented:
> What do you think of having different default values for the MVS is pytest mode? Specifically I mean "-f" and "-log warning". 

@smartie2076 - I though about that and I think it is better to use the parser as is and to pass it the "-f" and "-log warning" explicitly if wanted. Otherwise we might get confused with default values which change just for tests
